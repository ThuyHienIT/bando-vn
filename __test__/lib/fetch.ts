const mockFetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
    text: () => Promise.resolve(''),
    clone: () => ({
      json: () => Promise.resolve({}),
      text: () => Promise.resolve(''),
    }),
  })
);

if (!global.fetch) {
  (global as any).fetch = mockFetch;
}

export { mockFetch };
