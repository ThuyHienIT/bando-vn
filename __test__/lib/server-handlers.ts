// this is put into here so I can share these same handlers between my tests

const handlers = [] as any;

export { handlers };
