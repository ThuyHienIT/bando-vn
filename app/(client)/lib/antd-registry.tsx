'use client';

// https://github.com/vercel/next.js/issues/44125#issuecomment-1372270391
import { useServerInsertedHTML } from 'next/navigation';
import React, { useState } from 'react';

import { createCache, extractStyle, StyleProvider } from '@ant-design/cssinjs';

export function AntdStyleRegistry({ children }: { children: React.ReactNode }) {
  const [cache] = useState(() => createCache());

  const render = <>{children}</>;

  useServerInsertedHTML(() => {
    return (
      <script
        dangerouslySetInnerHTML={{
          __html: `</script>${extractStyle(cache)}<script>`,
        }}
      />
    );
  });

  if (typeof window !== 'undefined') {
    return render;
  }

  return <StyleProvider cache={cache}>{render}</StyleProvider>;
}
