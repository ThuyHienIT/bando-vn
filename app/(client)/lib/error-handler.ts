export async function errorHandler<T>(
  fn: Function,
  defaultValue: T
): Promise<T> {
  try {
    return await fn();
  } catch (e: any) {
    return Promise.resolve(defaultValue);
  }
}
