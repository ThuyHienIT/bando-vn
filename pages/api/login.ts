import { getIronSession } from 'iron-session';
import { NextApiRequest, NextApiResponse } from 'next';

import { ironOptions } from '@lib/config';
import { tryParseJson } from '@lib/tryParseJSON';
import { loginUser } from '@models/user';

async function routeHandler(req: NextApiRequest, res: NextApiResponse) {
  try {
    if (req.method === 'POST') return get(req, res);

    return res.status(405).json({ message: 'Method Not Allowed' });
  } catch (e: any) {
    console.debug('/api/login/ ERROR::', e);
    return res
      .status(500)
      .json({ message: 'Something went wrong. Please try again later.' });
  }
}

async function get(req: NextApiRequest, res: NextApiResponse) {
  const payload = tryParseJson(req.body);
  if (!payload.email || !payload.password)
    return res.status(400).json({ message: 'Email & Password are required' });

  const session = await getIronSession(req, res, ironOptions);
  const user = await loginUser(payload.email, payload.password);

  if (!user)
    return res.status(400).json({ message: 'Email or Password is invalid' });

  // get user from database then:
  session.user = user;
  req.session = session;

  await req.session.save();

  return res.status(200).json(user);
}
export default routeHandler;
