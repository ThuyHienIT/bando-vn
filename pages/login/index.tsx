import { getIronSession } from 'iron-session';
import { GetServerSidePropsContext, NextPage } from 'next';
import { redirect } from 'next/navigation';

import { ironOptions } from '@lib/config';
import LoginPage from '@pages/LoginPage';

const Page: NextPage<any> = (props) => {
  return <LoginPage />;
};

export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
  const session = await getIronSession(ctx.req, ctx.res, ironOptions);

  if (session.user)
    return {
      redirect: {
        destination: '/admin',
      },
    };

  return { props: {} };
};

export default Page;
