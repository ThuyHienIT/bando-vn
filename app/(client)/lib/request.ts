import { RequestError } from 'lib/errorClasses';

const PREFIX = 'http://localhost:3000';
export default async function request(url: string, options?: RequestInit) {
  const token = '';
  let { headers: optHeaders, ...restOpts } = options || {};
  const fullUrl = url.startsWith('http') ? url : `${PREFIX}${url}`;

  const headers: Record<string, any> = {
    'Content-Type': 'application/json',
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
    ...optHeaders,
  };

  if (headers['Content-Type'] === 'remove') {
    delete headers['Content-Type'];
  }

  const resp = await fetch(fullUrl, {
    method: 'GET',
    headers,
    ...restOpts,
  });

  // Error
  if (resp.status < 200 || resp.status >= 300) {
    throw new RequestError(resp.status, (await resp.json())?.message);
  }

  const customHeader = (headers as any)?.['Content-Type'];

  if (customHeader && customHeader.startsWith('text'))
    return resp.clone().text();
  if (customHeader && customHeader === 'application/json')
    return resp.clone().json();

  // Success
  return resp;
}
