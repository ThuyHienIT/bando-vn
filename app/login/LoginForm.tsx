'use client';
import { Button, Card, Form, Input, notification, Typography } from 'antd';
import request from 'app/(client)/lib/request';
import Image from 'next/image';
import { useRouter } from 'next/navigation';
import React, { memo, useCallback, useState } from 'react';
import styled from 'styled-components';

interface Props {}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  background-image: url('https://picsum.photos/1084/992/?blur');
  background-size: cover;
  background-position: center center;
  min-height: 100vh;
  background-position: center;
  position: relative;
  padding: 1rem;

  &::before {
    content: '';
    background-color: rgba(0, 0, 0, 0.7);
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;
  }
`;

const CardStyle = styled(Card)`
  border-radius: 4px;
  box-shadow: 4px 4px 5px rgba(0, 0, 0, 0.3);
  max-width: 500px;
  width: 100%;
`;

const Logo = styled.div`
  display: flex;
  align-items: center;
  gap: 8px;
  margin-bottom: 16px;
`;

export const LoginForm = memo(function LoginForm(props: Props) {
  const [form] = Form.useForm();
  const [loading, setLoading] = useState(false);
  const email = Form.useWatch('email', form);
  const password = Form.useWatch('password', form);
  const [api, contextHandler] = notification.useNotification();
  const router = useRouter();

  const handleSubmit = useCallback(
    async (values: { email: string; password: string }) => {
      try {
        setLoading(true);
        await request('/api/login', {
          method: 'POST',
          body: JSON.stringify({
            email: values.email,
            password: values.password,
          }),
        });

        router.push('/');
      } catch (e: any) {
        api.error({ message: e.message });
      } finally {
        setLoading(false);
      }
    },
    [api, router]
  );

  return (
    <Wrapper>
      {contextHandler}
      <CardStyle>
        <Logo>
          <Image
            src="/logo.png"
            alt="Facility Booking"
            width={50}
            height={50}
          />
          <Typography.Title level={4} style={{ marginBottom: 0 }}>
            Facility Booking
          </Typography.Title>
        </Logo>
        <Form form={form} onFinish={handleSubmit} layout="vertical">
          <Typography.Paragraph type="secondary">
            Enter your Email Address and Password to login
          </Typography.Paragraph>

          <Form.Item
            label="Email"
            name="email"
            rules={[{ required: true, type: 'email' }]}
          >
            <Input placeholder="abc@example.com" type="email" allowClear />
          </Form.Item>

          <Form.Item
            label="Password"
            name="password"
            rules={[{ required: true }]}
          >
            <Input type="password" allowClear />
          </Form.Item>

          <Button
            type="primary"
            htmlType="submit"
            loading={loading}
            disabled={!email || !password}
          >
            Login
          </Button>
        </Form>
      </CardStyle>
    </Wrapper>
  );
});
