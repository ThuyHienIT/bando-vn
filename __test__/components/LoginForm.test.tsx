import { mockFetch } from '__test__/lib/fetch';
import { LoginForm } from 'app/login/LoginForm';

import { fireEvent, render, waitFor } from '@testing-library/react';

beforeAll(() => {
  mockFetch.mockClear();
});

describe('LoginForm', () => {
  it('renders LoginForm unchanged', () => {
    const { container } = render(<LoginForm />);

    expect(container).toMatchSnapshot();
  });

  it('submit with valid data', async () => {
    const { container } = render(<LoginForm />);

    const emailInput = container.querySelector('[id="email"]');
    const passwordInput = container.querySelector('[id="password"]');
    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    if (emailInput)
      fireEvent.change(emailInput, {
        target: { value: 'kqthang1505@gmail.com' },
      });
    if (passwordInput)
      fireEvent.change(passwordInput, { target: { value: '1234' } });

    const submitBtn = container.querySelector('[type="submit"]');

    expect(submitBtn).toBeInTheDocument();
    if (submitBtn) fireEvent.click(submitBtn);

    await waitFor(() => {
      expect((global as any).mockRouterPush).toBeCalled();
    });
  });

  it('submit with missing data', async () => {
    const { container } = render(<LoginForm />);

    const emailInput = container.querySelector('[id="email"]');
    const passwordInput = container.querySelector('[id="password"]');
    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    if (emailInput)
      fireEvent.change(emailInput, {
        target: { value: 'kqthang1505@gmail.com' },
      });

    const submitBtn = container.querySelector('[type="submit"]');

    expect(submitBtn).toBeInTheDocument();
    if (submitBtn) fireEvent.click(submitBtn);

    await waitFor(() => {
      expect((global as any).mockRouterPush).toBeCalled();
    });
  });
});
