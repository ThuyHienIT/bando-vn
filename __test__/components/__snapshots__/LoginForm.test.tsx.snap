// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders LoginForm unchanged 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-image: url('https://picsum.photos/1084/992/?blur');
  background-size: cover;
  background-position: center center;
  min-height: 100vh;
  background-position: center;
  position: relative;
  padding: 1rem;
}

.c0::before {
  content: '';
  background-color: rgba(0,0,0,0.7);
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
}

.c1 {
  border-radius: 4px;
  box-shadow: 4px 4px 5px rgba(0,0,0,0.3);
  max-width: 500px;
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  margin-bottom: 16px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="ant-card ant-card-bordered c1 css-dev-only-do-not-override-diro6f"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="c2"
        >
          <img
            alt="Facility Booking"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=%2Flogo.png&w=128&q=75"
            srcset="/_next/image?url=%2Flogo.png&w=64&q=75 1x, /_next/image?url=%2Flogo.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
          <h4
            class="ant-typography css-dev-only-do-not-override-diro6f"
            style="margin-bottom: 0px;"
          >
            Facility Booking
          </h4>
        </div>
        <form
          class="ant-form ant-form-vertical css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-typography ant-typography-secondary css-dev-only-do-not-override-diro6f"
          >
            Enter your Email Address and Password to login
          </div>
          <div
            class="ant-form-item css-dev-only-do-not-override-diro6f"
          >
            <div
              class="ant-row ant-form-item-row css-dev-only-do-not-override-diro6f"
            >
              <div
                class="ant-col ant-form-item-label css-dev-only-do-not-override-diro6f"
              >
                <label
                  class="ant-form-item-required"
                  for="email"
                  title="Email"
                >
                  Email
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control css-dev-only-do-not-override-diro6f"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <span
                      class="ant-input-affix-wrapper css-dev-only-do-not-override-diro6f"
                    >
                      <input
                        aria-required="true"
                        class="ant-input css-dev-only-do-not-override-diro6f"
                        id="email"
                        placeholder="abc@example.com"
                        type="email"
                        value=""
                      />
                      <span
                        class="ant-input-suffix"
                      >
                        <span
                          class="ant-input-clear-icon ant-input-clear-icon-hidden"
                          role="button"
                          tabindex="-1"
                        >
                          <span
                            aria-label="close-circle"
                            class="anticon anticon-close-circle"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-form-item css-dev-only-do-not-override-diro6f"
          >
            <div
              class="ant-row ant-form-item-row css-dev-only-do-not-override-diro6f"
            >
              <div
                class="ant-col ant-form-item-label css-dev-only-do-not-override-diro6f"
              >
                <label
                  class="ant-form-item-required"
                  for="password"
                  title="Password"
                >
                  Password
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control css-dev-only-do-not-override-diro6f"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <span
                      class="ant-input-affix-wrapper css-dev-only-do-not-override-diro6f"
                    >
                      <input
                        aria-required="true"
                        class="ant-input css-dev-only-do-not-override-diro6f"
                        id="password"
                        type="password"
                        value=""
                      />
                      <span
                        class="ant-input-suffix"
                      >
                        <span
                          class="ant-input-clear-icon ant-input-clear-icon-hidden"
                          role="button"
                          tabindex="-1"
                        >
                          <span
                            aria-label="close-circle"
                            class="anticon anticon-close-circle"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="close-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-primary"
            disabled=""
            type="submit"
          >
            <span>
              Login
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
